contrib/storage-phoenix/src/main/java/org/apache/drill/exec/store/phoenix/PhoenixBatchReader.java

```diff
@@ -67,42 +74,55 @@
 
   public PhoenixBatchReader(PhoenixSubScan subScan) {
     this.subScan = subScan;
+    this.impersonationEnabled = subScan.getPlugin().getContext().getConfig().getBoolean(ExecConstants.IMPERSONATION_ENABLED);
   }
 
   @Override
   public boolean open(SchemaNegotiator negotiator) {
+    return impersonationEnabled
+      ? ugi.doAs((PrivilegedAction<Boolean>) () -> processOpen(negotiator))
+      : processOpen(negotiator);
+  }
+
+  private boolean processOpen(SchemaNegotiator negotiator) {
     try {
       errorContext = negotiator.parentErrorContext();
-      conn = subScan.getPlugin().getDataSource().getConnection();
-      pstmt = conn.prepareStatement(subScan.getSql());
+      DataSource ds = subScan.getPlugin().getDataSource(negotiator.userName());
+      PreparedStatement pstmt = ds.getConnection().prepareStatement(subScan.getSql());
       results = pstmt.executeQuery();
       meta = pstmt.getMetaData();
     } catch (SQLException e) {
       throw UserException
-              .dataReadError(e)
-              .message("Failed to execute the phoenix sql query. " + e.getMessage())
-              .addContext(errorContext)
-              .build(logger);
+        .dataReadError(e)
+        .message("Failed to execute the phoenix sql query. " + e.getMessage())
+        .addContext(errorContext)
+        .build(logger);
     }
```

Comment:
```suggestion
    } finally {
    pstmt.close();
```

Something flagged by LGTM.